gap/digraph.gi

```diff
@@ -1368,6 +1423,259 @@ function(filt, n, p)
   return DigraphNC(IsMutableDigraph, RANDOM_DIGRAPH(n, Int(p * 10000)));
 end);
 
+BindGlobal("DIGRAPHS_FillOutGraph", function(n, p, adjacencyList)
+
+    local i, j;
+
+    for i in [1 .. n] do
+        for j in [1 .. n] do
+            if (not (j in adjacencyList[i])) then
+                 if Float(Random([0 .. 99])/100) < Float(p) then
+                    Add(adjacencyList[i], j);
+                 fi;
+            fi;
+        od;
+    od;
+
+    return adjacencyList;
+
+end);
+
+InstallMethod(RandomDigraphCons,
+"for IsHamiltonianDigraph, a positive integer, and a float",
+[IsHamiltonianDigraph, IsPosInt, IsFloat],
+function(filt, n, p)
+    local adjacencyList, vertices, startVertex, hamiltonianCycle, x, i, j, graph;
+  
+    adjacencyList := [];
+  
+    for i in [1 .. n] do
+        Add(adjacencyList, []);
+    od;
+  
+    # Edge Case
+    if n = 1 then
+        if Float(Random([0 .. 99])/100) < Float(p) then
+            Add(adjacencyList[1], 1);
+            return Digraph(adjacencyList);
+        else
+            return Digraph(adjacencyList);
+        fi;
+    fi;
+  
+    vertices:= [1 .. n];
+    
+    # Starting from a random vertex, we create a Hamiltonian cycle
+    startVertex := Remove(vertices, Random(vertices));
+    hamiltonianCycle := [startVertex];
+    
+    # While there are remaining n-1 vertices to be added to the Hamiltonian cycle
+    for x in [1 .. n-1] do
+        # Create a random edge from the last vertex in the cycle, to random vertex not in the cycle
+        i := hamiltonianCycle[x];
+        j := Random(vertices);
+        Remove(vertices, Position(vertices, j));
```

Comment:
To avoid a search here, replace these two lines with:
```
j := Remove(vertices, Random([1 .. Length(vertices)]));
```
It looks worse, but the linear search is expensive.